README.md
# SocketIO native

Flutter socket.io-client using native platform.

## Getting Started

To use this plugin, add socketio_native as a dependency in your pubspec.yaml file.

## Android

Add the internet permissions to the AndroidManifest.xml.

Add **android:usesCleartextTraffic="true"** in application tag inside AndroidManifest.xml.

## iOS

iOS is currently not supported.

````dart
import 'package:flutter/material.dart';
import 'package:logger/logger.dart';
import 'package:socketio_native/socketio_native.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  late final SocketIO socketIO;

  Logger log = Logger(printer: PrettyPrinter(methodCount: 1));

  @override
  void initState() {
    super.initState();
    initSocket();
  }

  initSocket() async {
    try {
      Option option = Option();
      // option.setTransport([SocketIoTransport.polling, SocketIoTransport.websocket]);
      // option.setSecure(true);
      // option.setTimeout(60000);
      // option.setAuth({"api_key" : "20awCHxcod5NJ6Q8UFPuL7JjUHEdgm6BCT0oyZoo8Dl"});
      socketIO = await IO.create("http://192.168.1.183:3000", option: option);
      
      socketIO.onConnect((p0) {
        debugPrint("Connected>>>>");
        log.w({"CONNECTED": p0});
        socketIO.emit("message", "Hello");
      });
      socketIO.onDisconnect((p0) =>
      {
        log.w({"DISCONNECT": p0})
      });
      socketIO.onError((p0) =>
      {
        log.e({"ERROR": p0})
      });
      socketIO.on("message", (p0) {
        log.w({"Message": p0});
        socketIO.emit("message_ack", "Ok", ack: (ok) {
          log.w({"MessageACK": ok});
        });
      });
    } catch (e) {
      debugPrint(e.toString());
    }
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: const Center(
          child: Text('Running ...'),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            socketIO.emit("message", "message");
          },
          child: const Icon(Icons.add),
        ),
      ),
    );
  }
}

````

## Maintainer

[Tsiresy Mila](https://tsiresymila.sucthapp.com)

If you experience any problems with this package, please create an issue on Github. Pull requests
are also very welcome.



CHANGELOG.md
## 0.0.1

* FEATURE : SocketIO android work.

## 0.0.1+1

* CHANGE: Add licence

## 0.0.1+2

* CHANGE: Add description

## 0.0.1+3

* CHANGE: Update description

## 0.0.1+4

* Activate autoConnect

## 0.0.2

* Refactoring

## 0.0.2+1

* https working

## 0.0.2+2

* update readme
* 
## 0.0.3

* JSON toJson and fromJson using GSON

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:socketio_native_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data!.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/socketio_native_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:socketio_native/socketio_native.dart';
import 'package:socketio_native/socketio_native_platform_interface.dart';
import 'package:socketio_native/socketio_native_method_channel.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

class MockSocketIoNativePlatform
    with MockPlatformInterfaceMixin
    implements SocketIoNativePlatform {

  @override
  Future<String?> getPlatformVersion() => Future.value('42');

  @override
  Future<void> setUri(String uri) {
    // TODO: implement setUri
    throw UnimplementedError();
  }

  @override
  Future<void> setOption(String event, data) {
    // TODO: implement setOption
    throw UnimplementedError();
  }

  @override
  Future<void> callSocketIoMethod(String event) {
    throw UnimplementedError();
  }

  @override
  Future callSocketIoMethodWithCallback(String method, data, Map<String, dynamic>? other) {
    // TODO: implement callSocketIoMethodWithCallback
    throw UnimplementedError();
  }

  
}

void main() {
  final SocketIoNativePlatform initialPlatform = SocketIoNativePlatform.instance;

  test('$MethodChannelSocketIoNative is the default instance', () {
    expect(initialPlatform, isInstanceOf<MethodChannelSocketIoNative>());
  });

  test('initSocketIo', () async {
    SocketIO socketIo = await IO.create("https://realtimeio.sutchapp.com");
    MockSocketIoNativePlatform fakePlatform = MockSocketIoNativePlatform();
    SocketIoNativePlatform.instance = fakePlatform;
    expect(socketIo, isNotNull);
  });
}


test/socketio_native_method_channel_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:socketio_native/socketio_native_method_channel.dart';

void main() {
  MethodChannelSocketIoNative platform = MethodChannelSocketIoNative();
  const MethodChannel channel = MethodChannel('socketio_native');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await platform.getPlatformVersion(), '42');
  });
}


lib/socketio_native.dart
import 'dart:async';
import 'dart:convert';
import 'package:logger/logger.dart';
import 'package:socketio_native/socketio_native_event_channel.dart';
import 'socketio_native_platform_interface.dart';

// SocketIo transport
enum SocketIoTransport {
  websocket,
  polling,
}

/*

Initialized Option
 */
class Option {
  final Map<String, dynamic> _options = {
    "Option[setForceNew]": false,
    "Option[setMultiplex]": true,
    "Option[setUpgrade]": true,
    "Option[setRememberUpgrade]": false,
    "Option[setPath]": "/socket.io/",
    "Option[setQuery]": null,
    "Option[setExtraHeaders]": null,
    "Option[setReconnection]": true,
    "Option[setReconnectionDelay]": 1000,
    "Option[setReconnectionDelayMax]": 5000,
    "Option[setRandomizationFactor]": 0.5,
    "Option[setTimeout]": 20000,
    "Option[setTransport]": [
      SocketIoTransport.polling.name,
      SocketIoTransport.websocket.name
    ]
  };

  setForceNew(bool isForceNew) {
    _options["Option[setForceNew]"] = isForceNew;
  }

  setSecure(bool isSecure) {
    _options["Option[setSecure]"] = isSecure;
  }

  setReconnection(bool reconnect) {
    _options["Option[setReconnection]"] = reconnect;
  }

  setReconnectionDelay(double delay) {
    _options["Option[setReconnectionDelay]"] = double.tryParse(delay.toString());
  }

  setReconnectionDelayMax(double delay) {
    _options["Option[setReconnectionDelayMax]"] = double.tryParse(delay.toString());
  }

  setReconnectionAttempts(int attempts) {
    _options["Option[setReconnectionDelayMax]"] = attempts;
  }

  setMultiplex(bool isMultiplex) {
    _options["Option[setMultiplex]"] = isMultiplex;
  }

  setUpgrade(bool upgrade) {
    _options["Option[setUpgrade]"] = upgrade;
  }

  setRememberUpgrade(bool upgrade) {
    _options["Option[setRememberUpgrade]"] = upgrade;
  }

  setRandomizationFactor(double factor) {
    _options["Option[setRandomizationFactor]"] = factor;
  }

  setTimeout(double timeout) {
    _options["Option[setTimeout]"] = double.tryParse(timeout.toString());
  }

  disableAutoConnect() {
    _options["Option[disableAutoConnect]"] = true;
  }

  setTransport(List<SocketIoTransport> transports) {
    _options["Option[setTransport]"] = transports.map((e) => e.name).toList();
    toString();
  }

  setPath(String path) {
    _options["Option[setPath]"] = path;
  }

  setQuery(String query) {
    _options["Option[setQuery]"] = query;
  }

  setExtraHeaders(Map<String, List<String>> extras) {
    _options["Option[setExtraHeaders]"] = extras;
  }

  setAuth(Map<String, dynamic> auth) {
    _options["Option[setAuth]"] = auth;
  }

  // To Map data
  Map<String, dynamic> build() {
    return _options;
  }
}

class IO {
  static Future<SocketIO> create(String uri, {Option? option}) async {
    SocketIO io =  SocketIO(uri: uri, option: option);
    await io.init();
    return io;
  }
}

// SocketIo class
class SocketIO {
  final String uri;
  late final Option option;
  late Logger log;
  bool? connected;
  String? id;
  bool active = false;

  Map<String, StreamSubscription<dynamic>> channels = {};

  SocketIO({required this.uri, Option? option}) {
    this.option = option ?? Option();
    log = Logger(
        printer: PrettyPrinter(
      methodCount: 1,
      stackTraceBeginIndex: 1,
      errorMethodCount: 1,
    ));
  }

  // init socket
  init() async {
    channels = {};
    await SocketIoNativePlatform.instance.setUri(uri);
    for (MapEntry<String, dynamic> opt in option.build().entries) {
      await SocketIoNativePlatform.instance.setOption(opt.key, opt.value);
    }
    await SocketIoNativePlatform.instance.callSocketIoMethod("SocketIO[init]");
  }

  // reconnect socket manually
  reconnect() async {
    await SocketIoNativePlatform.instance
        .callSocketIoMethod("SocketIO[connect]");
  }

  // connect from socket manually
  connect() async {
    return await SocketIoNativePlatform.instance
        .callSocketIoMethod("SocketIO[connect]");
  }
  // disconnect from socket
  disconnect() async {
    return await SocketIoNativePlatform.instance
        .callSocketIoMethod("SocketIO[disconnect]");
  }
  // send data
  send(dynamic data) async {
    await SocketIoNativePlatform.instance
        .callSocketIoMethodWithCallback("SocketIO[send]", [data], null);
  }

  // emit data
  emit(String event, dynamic data, {Function(dynamic)? ack}) async {
    SocketIoNativePlatform.instance.callSocketIoMethodWithCallback(
        "SocketIO[emit]", [data], {"event": event}).then((value) {
      if (ack != null) {
        try{
          var data = jsonDecode(value.toString());
          ack.call(data);
        }catch(e){
          ack.call(value);
        }
      }
    }).onError((error, stackTrace) {
      log.e({"ERROR": error});
    });
  }

  // emit with timeout
  emitWithTimeout(String event, dynamic data, int timeout,
      {Function(dynamic)? ack}) async {
    try {
      SocketIoNativePlatform.instance.callSocketIoMethodWithCallback(
          "SocketIO[emitTimeout]",
          [data],
          {"event": event, "timeout": timeout}).then((value) {
        if (ack != null) {
          ack.call(value);
        }
      });
    } catch (e) {
      throw Exception("Timeout exception");
    }
  }

  // clear event listener
  off(String event) async {
    await SocketIoNativePlatform.instance
        .callSocketIoMethodWithCallback("SocketIO[off]", event, null);
  }

  // clear all event listener
  offAny() async {
    await SocketIoNativePlatform.instance
        .callSocketIoMethod("SocketIO[offAll]");
  }

  // listen on event
  on(String event, Function(dynamic) callback) {
    _createListener("SocketIO[on]", event, callback);
  }

  // listen single
  once(String event, Function(dynamic) callback) {
    _createListener("SocketIO[once]", event, callback);
  }

  // listen on connected
  onConnect(Function(dynamic) callback) {
    _createListener("SocketIO[onConnect]", "connect", (data) {
      connected = data["connected"] ?? false;
      id = data["id"];
      active = data["active"];
      callback(this);
    });
  }

  // listen error from socket
  onError(Function(dynamic) callback) {
    _createListener("SocketIO[onError]", "connect_error", callback);
  }

  // listen when socket disconnected
  onDisconnect(Function(dynamic) callback) {
    _createListener("SocketIO[onDisconnect]", "disconnect", (data) {
      connected = data["connected"];
      id = data["id"];
      active = data["active"];
      callback(this);
    });
  }

  // listen for any events
  onAny(Function(dynamic) callback) {
    SocketIoNativePlatform.instance
        .callSocketIoMethod("SocketIO[onAny]")
        .then((value) {
      EventChannelSocketIoNative(null)
          .eventChannel
          .receiveBroadcastStream()
          .listen((data) {
        try{
          var value = jsonDecode(data);
          callback(value);
        }catch(e){
          callback(data);
        }
      }, onError: (error) {
        log.e({"ERROR": error});
      }, onDone: () {
        log.w("DONNE");
      });
    });
  }

  // close socket
  close() {
    disconnect();
    for (var ch in channels.values) {
      ch.cancel();
    }
  }

  // private create listener
  _createListener(String method, String event, Function(dynamic) callback) {
    if (channels.keys.contains(event)) {
      channels[event]?.cancel();
    }
    SocketIoNativePlatform.instance
        .callSocketIoMethodWithCallback(method, event, null)
        .then((value) {
      EventChannelSocketIoNative chl = EventChannelSocketIoNative(event);
      StreamSubscription<dynamic> stream =
          chl.eventChannel.receiveBroadcastStream().listen((data) {
            try{
              var value = jsonDecode(data);
              callback(value);
            }catch(e){
              callback(data);
            }
      }, onError: (error) {
        log.e({"ERROR": error});
      }, onDone: () {
        log.w("DONNE");
      },cancelOnError: true);
      channels[event] = stream;

    });
  }
}


lib/socketio_native_method_channel.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

import 'socketio_native_platform_interface.dart';

/// An implementation of [SocketIONativePlatform] that uses method channels.
class MethodChannelSocketIoNative extends SocketIoNativePlatform {
  /// The method channel used to interact with the native platform.
  @visibleForTesting
  final methodChannel = const MethodChannel('socketio_native');

  @override
  Future<String?> getPlatformVersion() async {
    final version =
        await methodChannel.invokeMethod<String>('getPlatformVersion');
    return version;
  }

  @override
  Future<dynamic> setUri(String uri) async {
    return await methodChannel
        .invokeMethod<dynamic>("SocketIO[setUri]", {"data": uri});
  }

  @override
  Future<dynamic> setOption(String method, data) async {
    return await methodChannel.invokeMethod<dynamic>(method, {"data": data});
  }

  @override
  Future<dynamic> callSocketIoMethod(String method) async {
    return await methodChannel.invokeMethod<dynamic>(method);
  }

  @override
  Future<dynamic> callSocketIoMethodWithCallback(
      String method, dynamic data, Map<String, dynamic>? other) async {
    return await methodChannel
        .invokeMethod<dynamic>(method, {"data": data, ...(other ?? {})});
  }
}


lib/socketio_native_platform_interface.dart
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

import 'socketio_native_method_channel.dart';

abstract class SocketIoNativePlatform extends PlatformInterface {
  /// Constructs a SocketIoNativePlatform.
  SocketIoNativePlatform() : super(token: _token);

  static final Object _token = Object();

  static SocketIoNativePlatform _instance = MethodChannelSocketIoNative();

  /// The default instance of [SocketIoNativePlatform] to use.
  ///
  /// Defaults to [MethodChannelSocketIONative].
  static SocketIoNativePlatform get instance => _instance;

  /// Platform-specific implementations should set this with their own
  /// platform-specific class that extends [SocketIoNativePlatform] when
  /// they register themselves.
  static set instance(SocketIoNativePlatform instance) {
    PlatformInterface.verifyToken(instance, _token);
    _instance = instance;
  }

  Future<String?> getPlatformVersion() {
    throw UnimplementedError('platformVersion() has not been implemented.');
  }

  Future<void> setUri(String uri) {
    throw UnimplementedError('setUri() has not been implemented.');
  }

  Future<void> setOption(String method, dynamic data) {
    throw UnimplementedError('setOption() has not been implemented.');
  }

  Future<void> callSocketIoMethod(String method) {
    throw UnimplementedError('setOption() has not been implemented.');
  }

  Future<dynamic> callSocketIoMethodWithCallback(
      String method, dynamic data, Map<String, dynamic>? other) {
    throw UnimplementedError('setOption() has not been implemented.');
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:logger/logger.dart';
import 'package:socketio_native/socketio_native.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatefulWidget {
  const MyApp({super.key});

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  late final SocketIO socketIO;

  Logger log = Logger(printer: PrettyPrinter(methodCount: 1));

  @override
  void initState() {
    super.initState();
    initSocket();
  }

  initSocket() async {
    try {
      Option option = Option();
      option.setTransport([SocketIoTransport.polling,SocketIoTransport.websocket]);
      option.setSecure(true);
      option.setTimeout(60000);
      option.setAuth({"api_key" : "4Mz8zGuOPFNOi95QMhjjZ85XsQIZzXBQ5VLJY6VDuS26"});
      socketIO = await IO.create("http://192.168.88.6:3000", option: option);
      socketIO.onConnect((p0) {
        debugPrint("Connected>>>>");
        log.w({"CONNECTED": p0});
        socketIO.emit("message", {"data": "Hello"});
      });
      socketIO.onDisconnect((p0) => {
        log.w({"DISCONNECT": p0})
      });
      socketIO.onError((p0) => {
        log.e({"ERROR": p0})
      });
      socketIO.on("message", (p0) {
        log.w({"Message": p0});
        socketIO.emit("message_ack", "Ok", ack: (ok) {
          log.w({"MessageACK": ok});
        });
      });
      // await socketIO.connect();
    } catch (e) {
      log.e(e.toString());
    }
  }

  @override
  void dispose() {
    socketIO.close();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: const Center(
          child: Text('Running ...'),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            socketIO.emit("message", "message");
          },
          child: const Icon(Icons.add),
        ),
      ),
    );
  }
}
